### PR TITLE
Testing chapters

### PR DESCRIPTION
**Pull request by serejkus**
Solves #111.

I've made all examples via rust,ignore annotation since rust-by-example requires `main` function, but testing is focused on library crates.

### CODE REVIEW

File: src/testing.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Testing
+
+Rust is a programming language that cares a lot about correctness and it
+includes support for writing software tests within the language itself.
+
+Testing comes in three "tastes":
```
Comment (steveklabnik) [inline src/testing.md]: Could we say "styles" instead of "tastes"?

File: src/testing.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+* [Unit][unit] testing.
+* [Doc][doc] testing.
+* [Integration][integration] testing.
+
+Also Rust has support for spectifying additional dependencies for tests:
+* [Dev-dependencies][dev-dependencies]
```
Comment (steveklabnik) [inline src/testing.md]: a blank line above this one would be nice, give it some room to breathe

File: src/testing/doc_testing.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
+# Documentation testing
+
+The primary way of documenting a Rust project is through annotating the source
+code. Documentation comments are writen in [markdown][markdown] and support code
+blocks in them. Rust takes care about correctnes, so these code blocks are
+compiled and used as tests.
+
+```rust,ignore
+#![crate_name = "doccomments"]
```
Comment (steveklabnik) [inline src/testing/doc_testing.md]: this should be left out

File: src/testing/doc_testing.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+/// First line is a short summary describing function.
+///
+/// The next lineas present detailed documentation. Code blocks start with
```
Comment (steveklabnik) [inline src/testing/doc_testing.md]: you have an extra 'a' in 'lines'

File: src/testing/doc_testing.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+///
+/// ```
+/// let result = doccomments::add(2, 3);
+/// assert_eq!(result, 5);
+/// ```
+pub fn add(a: i32, b: i32) -> i32 {
+    a + b
+}
+
+/// Usually doc comments may include sections Examples, Panics and Failures.
```
Comment (steveklabnik) [inline src/testing/doc_testing.md]: > may include the sections "Examples," "Panics," and "Failures."

File: src/testing/doc_testing.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+///
+/// ```rust,should_panic
+/// // panics on division by zero
+/// doccomments::div(10, 0);
+/// ```
+pub fn div(a: i32, b: i32) -> i32 {
+    if b == 0 {
+        panic!("Divide-by-zero error");
+    }
+    a / b
```
Comment (steveklabnik) [inline src/testing/doc_testing.md]: could you add a blank line above this one please?

File: src/testing/doc_testing.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+test src/lib.rs - div (line 21) ... ok
+test src/lib.rs - div (line 31) ... ok
+
+test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
+```
+
+## Motivation behind documentation tests
+
+The main prupose of documentation tests is to serve as an examples that exercise
+the functionality. One of the most important
```
Comment (steveklabnik) [inline src/testing/doc_testing.md]: this sentence is a fragment, what about

> that exercise the functionality, which is one of the most important guidelines.

File: src/testing/doc_testing.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+[guidelines][question-instead-of-unwrap]. It allows using examples from docs as
+a complete code snippets. But using `?` makes compilation fail since `main`
```
Comment (steveklabnik) [inline src/testing/doc_testing.md]: no need for 'a' here

File: src/testing/doc_testing.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+returns `unit`. The ability to hide some source lines from documentation comes
+as a rescue: one may write `fn try_main() -> Result<(), ErrorType>`, hide it and
```
Comment (steveklabnik) [inline src/testing/doc_testing.md]: to the rescue

File: src/testing/integration_testing.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+and can test private code. Integration tests are etxernal to your crate and use
+only its public interface in the same way any other code would. Their purpose is
+to test that many parts of your library work correctly together.
+
+Cargo looks for integration tests in `tests` directory next to `src`.
+
+File `src/lib.rs`:
+
+```rust,ignore
+#![crate_name = "adder"]
```
Comment (steveklabnik) [inline src/testing/integration_testing.md]: this should be removed

File: src/testing/integration_testing.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+// Assume that crate is called adder, will have to extern it in integration test.
+pub fn add(a: i32, b: i32) -> i32 {
+    a + b
+}
+```
+
+File with test: `tests/integration_test.rs`:
+
+```rust,ignore
+// extern crate we're testing, same as any other code will do.
```
Comment (steveklabnik) [inline src/testing/integration_testing.md]: would do

File: src/testing/unit_testing.md [hunk 12]
```diff
@@ -0,0 +1,231 @@
+# Unit testing
+
+Tests are Rust functions that verify that the non-test code is functioning in
+the expected manner. The bodies of test functions typically perform some setup,
+run the code we want to test, then assert whether the results are what we
+expect.
+
+Typical scenario for grouping tests - adding `tests` [mod][mod] under with
```
Comment (steveklabnik) [inline src/testing/unit_testing.md]: I'd re-phrase this entirely:

> Most unit tests go into a `tests` [mod][mod] with the `#[cfg(test)]` [attribute]. Test functions are marked with the `#[test`] attribute.

File: src/testing/unit_testing.md [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+// This is a really bad adding function, its purpose is to fail in this
+// example.
+#[allow(dead_code)]
+fn bad_add(a: i32, b: i32) -> i32 {
+    a - b
+}
+
+#[cfg(test)]
+mod tests {
+    // Note as useful idiom: importing names from outer (for mod tests) scope.
```
Comment (steveklabnik) [inline src/testing/unit_testing.md]: > Note this useful idiom: using a glob to import names from the parent module.

File: src/testing/unit_testing.md [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+    tests::test_bad_add
+
+test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
+```
+
+## Testing panics
+
+To check functions that should panic under certain circumstances, use attribute
+`#[should_panic]`. This attribute accepts optional parameter `expected = ` with
+text of panic message: it helps making checks if function may panic in multiple
```
Comment (steveklabnik) [inline src/testing/unit_testing.md]: > with the text of the panic message. If your function can panic in multiple ways, it helps make sure your test is testing the correct panic.

File: src/testing/unit_testing.md [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+   Doc-tests tmp-test-should-panic
+
+running 0 tests
+
+test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
+```
+
+## Running specific tests
+
+To run specific tests one may specify test name to `cargo test` command.
```
Comment (steveklabnik) [inline src/testing/unit_testing.md]: > specify the test name

File: src/testing/unit_testing.md [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+   Doc-tests tmp-test-should-panic
+
+running 0 tests
+
+test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
+```
+
+## Ignoring tests
+
+Tests can be marked with `#[ignore]` attribute to exclude some tests. Or to run
```
Comment (steveklabnik) [inline src/testing/unit_testing.md]: > with the `#[ignore]` attribute

**Comment by serejkus** [comment]
Thanks for pointing those out. I don't have much spare time to fix these issues in the next few days, but I'll try to resolve them on a weekend.

**Comment by marioidival** [comment]
@steveklabnik this PR is good enough now?

**Comment by steveklabnik** [comment]
Yup, thanks!

**Pull request merged**